recipe/patches/0009-fix-missing-DLL-exports-in-vendored-upb.patch

```diff
@@ -0,0 +1,128 @@
+ 
+ target_include_directories(upb_mem_lib
+@@ -3689,6 +3695,12 @@ if(WIN32 AND MSVC)
+       DESTINATION ${gRPC_INSTALL_LIBDIR} OPTIONAL
+     )
+   endif()
++  # More details in third_party/upb/upb/port_def.inc;
++  # needs to be separate from gRPC_DLL_{IM,EX}PORTS
++  set_target_properties(upb_message_lib PROPERTIES DEFINE_SYMBOL "UPB_DLL_EXPORTS")
++  if(BUILD_SHARED_LIBS)
++    target_compile_definitions(upb_message_lib INTERFACE UPB_DLL_IMPORTS)
++  endif()
+ endif()
+ 
+ target_include_directories(upb_message_lib
+diff --git a/third_party/upb/upb/mem/alloc.c b/third_party/upb/upb/mem/alloc.c
+index d103f98f30..397be37907 100644
+--- a/third_party/upb/upb/mem/alloc.c
++++ b/third_party/upb/upb/mem/alloc.c
+@@ -24,4 +24,4 @@ static void* upb_global_allocfunc(upb_alloc* alloc, void* ptr, size_t oldsize,
+   }
+ }
+ 
+-upb_alloc upb_alloc_global = {&upb_global_allocfunc};
++UPB_DLL upb_alloc upb_alloc_global = {&upb_global_allocfunc};
+diff --git a/third_party/upb/upb/mem/alloc.h b/third_party/upb/upb/mem/alloc.h
+index 441731d2d8..5e743dedf3 100644
+--- a/third_party/upb/upb/mem/alloc.h
++++ b/third_party/upb/upb/mem/alloc.h
+@@ -52,7 +52,7 @@ UPB_INLINE void upb_free(upb_alloc* alloc, void* ptr) {
+ 
+ // The global allocator used by upb. Uses the standard malloc()/free().
+ 
+-extern upb_alloc upb_alloc_global;
++UPB_DLL extern upb_alloc upb_alloc_global;
+ 
+ /* Functions that hard-code the global malloc.
+  *
+diff --git a/third_party/upb/upb/mini_table/internal/message.c b/third_party/upb/upb/mini_table/internal/message.c
+index b25edacca7..05da188fc7 100644
+--- a/third_party/upb/upb/mini_table/internal/message.c
++++ b/third_party/upb/upb/mini_table/internal/message.c
+@@ -7,7 +7,7 @@
+ 
+ #include "upb/mini_table/internal/message.h"
+ 
+-const struct upb_MiniTable _kUpb_MiniTable_Empty = {
++UPB_DLL const struct upb_MiniTable _kUpb_MiniTable_Empty = {
+     .subs = NULL,
+     .fields = NULL,
+     .size = 0,
+diff --git a/third_party/upb/upb/mini_table/internal/message.h b/third_party/upb/upb/mini_table/internal/message.h
+index 3457ea2cd6..8fa3325911 100644
+--- a/third_party/upb/upb/mini_table/internal/message.h
++++ b/third_party/upb/upb/mini_table/internal/message.h
+@@ -65,7 +65,7 @@ extern "C" {
+ #endif
+ 
+ // A MiniTable for an empty message, used for unlinked sub-messages.
+-extern const struct upb_MiniTable _kUpb_MiniTable_Empty;
++UPB_DLL extern const struct upb_MiniTable _kUpb_MiniTable_Empty;
+ 
+ // Computes a bitmask in which the |l->required_count| lowest bits are set,
+ // except that we skip the lowest bit (because upb never uses hasbit 0).
+diff --git a/third_party/upb/upb/port/def.inc b/third_party/upb/upb/port/def.inc
+index f320821384..3f4d7e3cec 100644
+--- a/third_party/upb/upb/port/def.inc
++++ b/third_party/upb/upb/port/def.inc
+@@ -326,3 +326,30 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
+ #else
+ #define UPB_DESC(sym) google_protobuf_##sym
+ #endif
++
++// UPB_DLL
```

Comment:
Heroic stuff

Reply:
Haha, thanks! 🙏 

It's been a long ride indeed. We had been trying to switch to shared builds [since](https://github.com/conda-forge/grpc-cpp-feedstock/pull/246) ~1.48 at least, finally managing to do it with lots of patching [for](https://github.com/conda-forge/grpc-cpp-feedstock/pull/243) 1.55, and managing to upstream [something](https://github.com/grpc/grpc/pull/34103/files) for 1.59. It's a pity that support has since regressed, but at least we know how to fix it. 🙃